m0kv_metadata_parsing_tool/base64_encoder_decoder.cc

```diff
@@ -0,0 +1,193 @@
+  std::string ret;
+  int i = 0;
+  int j = 0;
+  unsigned char char_array_3[3];
+  unsigned char char_array_4[4];
+
+  while (in_len--) {
+    char_array_3[i++] = *(bytes_to_encode++);
+    if (i == 3) {
+      char_array_4[0] = (char_array_3[0] & 0xfc) >> 2;
+      char_array_4[1] = ((char_array_3[0] & 0x03) << 4) + ((char_array_3[1] & 0xf0) >> 4);
+      char_array_4[2] = ((char_array_3[1] & 0x0f) << 2) + ((char_array_3[2] & 0xc0) >> 6);
+      char_array_4[3] = char_array_3[2] & 0x3f;
+
+      for(i = 0; (i <4) ; i++)
+        ret += base64_chars[char_array_4[i]];
+      i = 0;
+    }
+  }
+
+  if (i)
+  {
+    for(j = i; j < 3; j++)
+      char_array_3[j] = '\0';
+
+    char_array_4[0] = (char_array_3[0] & 0xfc) >> 2;
+    char_array_4[1] = ((char_array_3[0] & 0x03) << 4) + ((char_array_3[1] & 0xf0) >> 4);
+    char_array_4[2] = ((char_array_3[1] & 0x0f) << 2) + ((char_array_3[2] & 0xc0) >> 6);
+    char_array_4[3] = char_array_3[2] & 0x3f;
+
+    for (j = 0; (j < i + 1); j++)
+      ret += base64_chars[char_array_4[j]];
+
+    while((i++ < 3))
+      ret += '=';
+
+  }
+
+  return ret;
+}
+
+/******************************************************************************
+ this function breaks the encoded string in hi and lo format and decodes string
+******************************************************************************/
+struct m0_uint128 to_m0_uint128(const std::string &id_str) {
+    m0_uint128 id = {0ULL, 0ULL};
+    std::size_t delim_pos = id_str.find("-");
+    if (delim_pos != std::string::npos) {
+    std::string dec_id_u_hi = base64_decode(id_str.substr(0, delim_pos));
+    std::string dec_id_u_lo = base64_decode(id_str.substr(delim_pos + 1));
+    if ((dec_id_u_hi.size() == sizeof(id.u_hi)) &&
+        (dec_id_u_lo.size() == sizeof(id.u_lo))) {
+        memcpy((void *)&id.u_hi, dec_id_u_hi.c_str(), sizeof(id.u_hi));
+        memcpy((void *)&id.u_lo, dec_id_u_lo.c_str(), sizeof(id.u_lo));
+    }
+    }
+    return id;
+}
+
+
+/*********************************************************************************
+ this function is used to divide the hex address in two parts and then encode them
+*********************************************************************************/
+std::string to_string(std::string input_string) {
+  std::size_t delim_pos = input_string.find(":");
+  std::string u_hi_hex = input_string.substr(0, delim_pos);
+  std::string u_lo_hex = input_string.substr(delim_pos + 1);
+  std::stringstream ss, ss1;
+  uint64_t u_hi, u_lo;
+  ss << u_hi_hex;
+  ss >> std::hex >> u_hi;
+  ss1 << u_lo_hex;
+  ss1 >> std::hex >> u_lo;
+  return base64_encode((unsigned char const *)&u_hi, sizeof(u_hi)) + "-" +
+         base64_encode((unsigned char const *)&u_lo, sizeof(u_lo));
+}
+
+/****************************************
+main function 
+****************************************/
+int main(int argc, char *argv[])
+{
+  if (std::string(argv[1]) == "-d") {
```

Comment:
What if I do not pass any argument ? In that case your argv array would be of size 1, so I guess accessing argv[1] would give core dump !!!

Check for size of argv array, if its size is greater than 1 or not.